### DISCUSSION TITLE
Pico PIO using an external clock.

### DISCUSSION THREAD

**Discussion by Phil-VK6PH** [micropython/micropython#14418] [RP2040 / Pico]
I've tested this code that sets the Pico PIO clock to 10 MHz, divides it by 10 then delivers 1 MHz to Pin 17. This works fine.

# 
import utime
from rp2 import PIO, asm_pio
from machine import Pin

@asm_pio(set_init=rp2.PIO.OUT_HIGH)
def divide10():
wrap_target()
set(pins,1) [4] # set output high for 5 cycles
set(pins,0) [4] # set output low for 5 cycles
wrap()

sm = rp2.StateMachine(0, divide10, freq = 10000000, set_base=Pin(17))

sm.active(1)

Instead of using the internal clock I'd like to use an external GPS locked 10 MHz clock connected to say Pin 16 to run the code. That way the 1 MHz clock would be very accurate and stable.

I've not been able to find any code that does this and would appreciate some assistance please.

**Comment by rkompass** [comment]
I think you need something like:
```py
@rp2.asm_pio(sideset_init=PIO.OUT_LOW)
def clock_divby10():
    wrap_target()
    wait(0, pin, 0)   .side(0)
    wait(1, pin, 0)   .side(0)
    wait(0, pin, 0)   .side(0)
    wait(1, pin, 0)   .side(0)
    wait(0, pin, 0)   .side(0)
    wait(1, pin, 0)   .side(0)
    wait(0, pin, 0)   .side(0)
    wait(1, pin, 0)   .side(0)
    wait(0, pin, 0)   .side(0)
    wait(1, pin, 0)   .side(0)
    wait(0, pin, 0)   .side(1)
    wait(1, pin, 0)   .side(1)    
    wait(0, pin, 0)   .side(1)
    wait(1, pin, 0)   .side(1)    
    wait(0, pin, 0)   .side(1)
    wait(1, pin, 0)   .side(1)    
    wait(0, pin, 0)   .side(1)
    wait(1, pin, 0)   .side(1)    
    wait(0, pin, 0)   .side(1)
    wait(1, pin, 0)   .side(1)    
    wrap()
```
Instead of `pin` also `gpio` may be used. The PIO clock should be at least 20MHz (higher is better), so it really is able to wait for the incoming GPS clock levels.
It might be I missed some argument for `@asm_pio`. You may check that yourself.

**Reply by peterhinch** [reply]
While the frequency will be accurate there will be phase jitter owing to the finite sampling frequency. As you say, a higher frequency is better in this regard. For ~zero phase noise ignore the Pico completely and use a TTL chip to divide by ten...

**Reply by Phil-VK6PH** [reply]
Thanks Peter.  I've built GPS locked 10 MHz VCXOs in the past using TTL and they have worked very well.  Now I'd like to try an alternative approach.  I share your concern regarding phase jitter, even when using a high frequency clock.  I was thinking that perhaps using the 10 MHz VCXO to clock the Pico could work.  It appears you can remove the crystal and connect a clock to the XIN pin.  However, if not using 12 MHz then the USB interface may not work.

**Comment by Phil-VK6PH** [comment]
I have found some code that enables a 10 MHz external clock to be connected to GPIO20 (pin 26). The code then divides this clock into different frequencies and provides them on GPIO pins. The code is here:

https://github.com/dorsic/PicoDIV?tab=readme-ov-file

This is written in C but a Micropython version would be easier for me to use. I'd appreciate if someone could indicate how this could be moved to Micropython.

**Comment by rkompass** [comment]
I suppose this is using the PWM peripheral on the RP2, which may be set up to count incoming pulses on a pin.
So a pin frequency may be used instead of the CPU clock to generate the PWM signal, including pre-division by a fractional clock divider (see RP2040 Datasheet page 529).
Together with the right choice of PWM TOP and compare levels a clock division by a desired number may be achieved.
Note however, that the registration of the incoming input pulse flanks occurs at regular CPU clock periods, so the jitter up to a full CPU clock period (8 or 4 ns, with 125 or 250 MHz CPU frequency) present at the PIO solution above is present here too.
If you want to try out this approach I recommend to study [this nice blog page](https://iosoft.blog/2023/08/21/picofreq_python/) which has a link to MP code examples.

**Comment by Phil-VK6PH** [comment]
Thank you for pointing out that blog page which is very interesting.  Since I'm building a GPS locked 10 MHz VCXO then I can't tolerate any jitter in the divided 10 MHz clock.  I have tried using a PWM to divide the Pico clock to 1 MHz and that appears to work OK.  The issue is that I can't find away to use an external clock to drive the PWM when using Micropython.
